src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs

```diff
@@ -1,51 +1,195 @@
+		#region Relinq adjusted code
 
-		internal NhPartialEvaluatingExpressionVisitor(ISessionFactoryImplementor sessionFactory)
+		/// <summary>
+		/// Takes an expression tree and finds and evaluates all its evaluatable subtrees.
+		/// </summary>
+		public static Expression EvaluateIndependentSubtrees(
+			Expression expressionTree,
+			PreTransformationParameters preTransformationParameters)
 		{
-			_sessionFactory = sessionFactory;
+			var partialEvaluationInfo = EvaluatableTreeFindingExpressionVisitor.Analyze(
+				expressionTree,
+				preTransformationParameters.EvaluatableExpressionFilter);
+			var visitor = new NhPartialEvaluatingExpressionVisitor(partialEvaluationInfo, preTransformationParameters);
+
+			return visitor.Visit(expressionTree);
 		}
 
+		// _partialEvaluationInfo contains a list of the expressions that are safe to be evaluated.
+		private readonly PartialEvaluationInfo _partialEvaluationInfo;
+		private readonly PreTransformationParameters _preTransformationParameters;
+
+		private NhPartialEvaluatingExpressionVisitor(
+			PartialEvaluationInfo partialEvaluationInfo,
+			PreTransformationParameters preTransformationParameters)
+		{
+			_partialEvaluationInfo = partialEvaluationInfo;
+			_preTransformationParameters = preTransformationParameters;
+		}
+
+		public override Expression Visit(Expression expression)
+		{
+			// Only evaluate expressions which do not use any of the surrounding parameter expressions. Don't evaluate
+			// lambda expressions (even if you could), we want to analyze those later on.
+			if (expression == null)
+				return null;
+
+			if (expression.NodeType == ExpressionType.Lambda || !_partialEvaluationInfo.IsEvaluatableExpression(expression))
+				return base.Visit(expression);
+
+			Expression evaluatedExpression;
+			try
+			{
+				evaluatedExpression = EvaluateSubtree(expression);
+			}
+			catch (Exception ex)
+			{
+				// Evaluation caused an exception. Skip evaluation of this expression and proceed as if it weren't evaluable.
+				var baseVisitedExpression = base.Visit(expression);
+
+				throw new HibernateException($"Evaluation failure on {baseVisitedExpression}", ex);
+			}
+
+			if (evaluatedExpression != expression)
+			{
+				evaluatedExpression = EvaluateIndependentSubtrees(evaluatedExpression, _preTransformationParameters);
+			}
+
+			#region NH additions
+
+			// When having multiple level closure, we have to evaluate each closure independently
+			if (evaluatedExpression is ConstantExpression constantExpression)
+			{
+				evaluatedExpression = VisitConstant(constantExpression);
+			}
+
+			// Variables in expressions are never a constant, they are encapsulated as fields of a compiler generated class.
+			// Skip detecting variables for DML queries as HQL does not support reusing parameters for them.
+			if (expression.NodeType != ExpressionType.Constant &&
+			    _preTransformationParameters.QueryMode == QueryMode.Select &&
```

Comment:
I wonder maybe it's better to introduce separate property for controlling parameters minimization (something like `_preTransformationParameters.MinimizeParameters` ) which by default is initialized based on `QueryMode` value.

Reply:
Probably better in case the information would be used elsewhere in the future, done.

Reply:
Good. But let's make it public. I see no point in hiding DTO like properties and it might be useful in some custom user code implementation. And comments regarding `Skip detecting variables for DML`  should be moved to `MinimizeParameters` initialization code (also should we move it right inside `PreTransformationParameters` ctor?)

Reply:
Made it public and moved the initialization inside ctor so that it can be set by the user.